server/src/main/java/io/druid/segment/realtime/RealtimeManager.java

```diff
@@ -139,49 +153,87 @@ public FireDepartmentMetrics getMetrics(String datasource)
+                    // Chaining query runners which wait on submitted chain query runners can make executor pools deadlock
+                    Iterables.transform(
+                        chiefsOfDataSource.values(), new Function<FireChief, QueryRunner<T>>()
+                        {
+                          @Override
+                          public QueryRunner<T> apply(FireChief input)
+                          {
+                            return input.getQueryRunner(query);
+                          }
+                        }
+                    )
+                )
+            );
+          }
+        }
+    );
+
+    return new UnionQueryRunner<>(factory.mergeRunners(MoreExecutors.sameThreadExecutor(), runners));
   }
 
   @Override
-  public <T> QueryRunner<T> getQueryRunnerForSegments(final Query<T> query, Iterable<SegmentDescriptor> specs)
+  public <T> QueryRunner<T> getQueryRunnerForSegments(final Query<T> query, final Iterable<SegmentDescriptor> specs)
   {
     final QueryRunnerFactory<T, Query<T>> factory = conglomerate.findFactory(query);
+    final List<QueryRunner<T>> runners = new ArrayList<>();
+    for (String dataSource : query.getDataSource().getNames()) {
+      final Map<Integer, FireChief> partitionChiefs = chiefs.get(dataSource);
+      if (partitionChiefs == null) {
+        continue;
+      }
 
-    Iterable<FireChief> chiefsOfDataSource = chiefs.get(Iterables.getOnlyElement(query.getDataSource().getNames()));
-    return chiefsOfDataSource == null ? new NoopQueryRunner() : factory.getToolchest().mergeResults(
-        factory.mergeRunners(
-            MoreExecutors.sameThreadExecutor(),
-            // Chaining query runners which wait on submitted chain query runners can make executor pools deadlock
-            Iterables.transform(
-                chiefsOfDataSource, new Function<FireChief, QueryRunner<T>>()
-                {
-                  @Override
-                  public QueryRunner<T> apply(FireChief input)
-                  {
-                    return input.getQueryRunner(query);
-                  }
-                }
-            )
-        )
-    );
+      final QueryToolChest<T, Query<T>> toolchest = factory.getToolchest();
+      final Iterable<QueryRunner<T>> subRunners = Iterables.transform(
+          specs,
+          new Function<SegmentDescriptor, QueryRunner<T>>()
+          {
+            @Nullable
+            @Override
+            public QueryRunner<T> apply(SegmentDescriptor spec)
+            {
+              FireChief retVal = partitionChiefs.get(spec.getPartitionNumber());
+              return retVal == null ? new NoopQueryRunner<T>() : retVal.getQueryRunner(query);
+            }
+          }
+      );
+      runners.add(
+          toolchest.mergeResults(factory.mergeRunners(MoreExecutors.sameThreadExecutor(), subRunners))
+      );
+    }
+    return new UnionQueryRunner<>(factory.mergeRunners(MoreExecutors.sameThreadExecutor(), runners));
```

Comment:
@nishantmonu51 What if a query's datasource is QueryDataSource, and its nested query's datasource is UnionDataSource?